resources/todomvc/architecture-examples/react-redux/src/components/main.jsx

```diff
@@ -0,0 +1,50 @@
+import { Component } from "react";
+import PropTypes from "prop-types";
+import { withRouter } from "react-router-dom";
+import Item from "./item";
+import Footer from "./footer";
+import { SHOW_ALL, SHOW_COMPLETED, SHOW_ACTIVE } from "../constants/todo-filters";
+
+const TODO_FILTERS = {
+    [SHOW_ALL]: () => true,
+    [SHOW_ACTIVE]: (todo) => !todo.completed,
+    [SHOW_COMPLETED]: (todo) => todo.completed,
+};
+
+export class Main extends Component {
+    static propTypes = {
+        todos: PropTypes.array.isRequired,
+        actions: PropTypes.object.isRequired,
+        location: PropTypes.object.isRequired,
+    };
+
+    render() {
+        const { todos, actions, location } = this.props;
+
+        const filteredTodos = todos.filter(TODO_FILTERS[location.pathname]);
+        const completedCount = todos.reduce((count, todo) => (todo.completed ? count + 1 : count), 0);
```

Comment:
So we're not memoizing here. Would it make sense to memoize using selectors, so that this is more redux-y?

We'd need a component that would listen for location changes and update the redux state with this info, so that the selector could pick it up.

Then with these selectors we could connect more components. For example `Footer` could be connected and get `completedCount` from the selector (as well as the `clearCompleted` action).

Reply:
BTW, by memoizing using selectors, I mean using the `reselect` library, like most apps do. (@redux/toolkit uses `reselect` under the hood for example)

Reply:
I tried to keep the previous implementation intact and therefore tried to avoid making changes the way redux was implemented. We could add another redux version of the todo app that uses a more modern approach and with hooks / functional components maybe?

Reply:
That said, I understand that the existing version in the tree doesn't do that either. So I'm OK if you want to discuss this topic first in the group, and do it in a follow-up then.

Reply:
@julienw - let's make things better! How about this approach:
* I'll update this version and use connected components closer where the todos / dispatch is used
* I'll add some selectors (and look into reselect) 
We just want to ensure that the performance doesn't degrade in any way with these updates.

Additionally, we could discuss adding a "modern" redux version that uses the currently recommended approach with @reduxjs/toolkit / functional components / hooks.

My personal goal for this version was to use a more "classic" version of redux in combination with React 16 and using class components / hocs, ect...

Reply:
I like this plan! Maybe just React 17 instead of 16? There's not much difference between 16 and 17 though.
I think we definitely need at least one class component-based implementation, so this is totally fine for me.
About the "modern" redux version my understanding is that we'll need to trade one benchmark for one, because we don't want to add more. But if we have good arguments in favor of this new version, maybe we'll get a pass ;-)

Call me radical :-) but TBH I don't understand this concern of "ensuring that the performance doesn't degrade", as long as the benchmark is close to what we see in the wild. If the performance degrades then this would mean that the real-life websites don't work performantly, and this needs to be fixed in these browsers.

Reply:
>  Additionally, the todomvc-react version is using 17.0.2. It's nice to have two different versions as workloads that represent what's being used.

I agree with this reasoning - if 16 is more commonly used with the patterns used in this workload makes sense to stick with it.